design/metadata-handling.md

```diff
@@ -0,0 +1,750 @@
+- **local-hostname**: The name of the BareMetalHost
+- **local_hostname**: The name of the BareMetalHost
+
+However, setting any of those values in the metaData secret will override those
+default values.
+
+### Metal3Machine changes
+
+The Metal3Machine will be modified as follow:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3Machine
+metadata:
+  name: machine-1
+  namespace: default
+spec:
+  dataTemplate:
+    name: nodepool-1
+    namespace: default
+status:
+  renderedData:
+    name: nodepool-1-0
+    namespace: default
+  metaData:
+    name: machine-1-metadata
+    namespace: default
+  networkData:
+    name: machine-1-networkdata
+    namespace: default
+```
+
+or alternatively
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3Machine
+metadata:
+  name: machine-1
+  namespace: default
+spec:
+  metaData:
+    name: machine-1-metadata
+    namespace: default
+  networkData:
+    name: machine-1-networkData
+    namespace: default
+status:
+  metaData:
+    name: machine-1-metadata
+    namespace: default
+  networkData:
+    name: machine-1-networkData
+    namespace: default
+```
+
+A `metaData` field would be added in the spec and in the status, consisting of a
+secret reference containing the name and the namespace of the secret containing
+the metaData for this Metal3Machine. A `networkData` field would be added to the
+spec and to the status objects, consisting of a secret
+reference containing the name and the namespace of the secret containing the
+network configuration for this Metal3Machine.
+
+The secret containing the metaData or the network data could be provided by the
+user directly using the `metaData` or `networkData` fields in the spec of the
+Metal3Machine.
+
+The `metaData` and `networkData` field in the `spec` section are for the user
+to give directly a secret to use as metaData or networkData. The `userData`,
+`metaData` and `networkData` fields in the `status` section are for the
+controller to store the reference to the secret that is actually being used,
+whether it is from one of the spec fields, or somehow generated. This is aimed
+at making a clear difference between the desired state from the user (whether
+it is with a DataTemplate reference, or direct `metaData` or `userData` secrets)
+and what the controller is actually using.
+
+A `dataTemplate` field would be added, consisting of an object reference
+to a Metal3DataTemplate object containing the templates for
+the metadata and network data generation for this Metal3Machine.
+A `renderedData` field will be added in the status and will be a reference to
+the Metal3Data object created for this machine. If the dataTemplate field is set
+but either the `renderedData`, `metaData` or `networkData` fields in the status
+are unset, then the Metal3Machine controller will wait until it can find the
+Metal3Data object and the rendered secrets. It will then populate those fields.
+
```

Comment:
It's not clear what happened if dataTemplate, metaData and networkData are set together. Which one has higher priority?